lib/fir/statements.h

```diff
@@ -0,0 +1,231 @@
+using PathVariable = const parser::Variable;
+
+struct Evaluation
+  : public SumTypeCopyMixin<
+        std::variant<Expression *, Variable *, PathVariable *>> {
+  SUM_TYPE_COPY_MIXIN(Evaluation)
+  Evaluation(PathVariable *pv) : SumTypeCopyMixin{pv};
+  template<typename A> Evaluation(A *a) : SumTypeCopyMixin{a} {}
+  std::string dump() const;
+};
+
+struct Stmt_impl {
+  using StatementTrait = std::true_type;
+};
+
+struct TerminatorStmt_impl : public Stmt_impl {
+  virtual std::list<BasicBlock *> succ_blocks() const;
+  using TerminatorTrait = std::true_type;
+};
+
+struct ReturnStmt : public TerminatorStmt_impl {
+  ~ReturnStmt() = default;
+  static ReturnStmt *Create();
+  static ReturnStmt *Create(Expression *expression);
+
+private:
+  Expression *returnValue;
+  explicit ReturnStmt(Expression *);
+};
+
+struct BranchStmt : public TerminatorStmt_impl {
+  friend Statement;
+  ~BranchStmt() = default;
+  static BranchStmt *Create(
+      Expression *condition, BasicBlock *trueBlock, BasicBlock *falseBlock);
+  static BranchStmt *Create(BasicBlock *succ);
+  bool hasCondition() const;
+  std::list<BasicBlock *> succ_blocks() const override;
+
+private:
+  explicit BranchStmt(
+      Expression *condition, BasicBlock *trueBlock, BasicBlock *falseBlock);
+  static constexpr unsigned TrueIndex{0u};
+  static constexpr unsigned FalseIndex{1u};
+  Expression *condition;
+  BasicBlock *succs[2];
+};
+
+struct SwitchStmt : public TerminatorStmt_impl {
+  using ValueSuccPairType = std::pair<Expression *, BasicBlock *>;
+  using ValueSuccPairListType = std::vector<ValueSuccPairType>;
+  ~SwitchStmt() = default;
+  static SwitchStmt *Create(const Evaluation &switchEval,
+      BasicBlock *defaultBlock, const ValueSuccPairListType &args);
+  BasicBlock *defaultSucc() const;
+  std::list<BasicBlock *> succ_blocks() const override;
+  const Evaluation &getCond() const;
+
+private:
+  explicit SwitchStmt(const Evaluation &condition, BasicBlock *defaultBlock,
+      const ValueSuccPairListType &args);
+  Evaluation condition;
+  ValueSuccPairListType valueSuccPairs;
+};
+
+struct IndirectBrStmt : public TerminatorStmt_impl {
+  using TargetListType = std::vector<BasicBlock *>;
+  ~IndirectBrStmt() = default;
+  static IndirectBrStmt *Create(
+      Variable *variable, const TargetListType &potentialTargets);
+
+private:
+  explicit IndirectBrStmt(
+      Variable *variable, const TargetListType &potentialTargets)
+    : variable{variable}, potentialTargets{potentialTargets} {}
+  Variable *variable;
+  TargetListType potentialTargets;
+};
+
+struct UnreachableStmt : public TerminatorStmt_impl {
+  static UnreachableStmt *Create();
+  ~UnreachableStmt() = default;
+
+private:
+  explicit UnreachableStmt() = default;
+};
+
+struct ActionStmt_impl : public Stmt_impl {
+  using ActionTrait = std::true_type;
+  // TODO: DynamicType is a placeholder for now
+  std::optional<evaluate::DynamicType> type_{std::nullopt};
+};
+
+struct AssignmentStmt
```

Comment:
It would be great if assignments could be templated on their types like expressions are.